lxd/db/deployments.go

```diff
@@ -0,0 +1,1135 @@
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	if rowsAffected <= 0 {
+		return api.StatusErrorf(http.StatusNotFound, "DeploymentShape not found")
+	}
+
+	_, err = c.tx.ExecContext(ctx, "DELETE FROM deployment_shape_configs WHERE deployment_shape_id=?", deploymentShapeID)
+	if err != nil {
+		return err
+	}
+
+	err = deploymentShapeConfigAdd(ctx, c.tx, deploymentShapeID, info.Config)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DeleteDeploymentShape deletes an existing deployment shape.
+func (c *ClusterTx) DeleteDeploymentShape(ctx context.Context, deploymentID int64, deploymentShapeID int64) error {
+	// Delete existing shape record.
+	res, err := c.tx.ExecContext(ctx, `
+			DELETE FROM deployment_shapes
+			WHERE deployment_id = ? and id = ?
+		`, deploymentID, deploymentShapeID)
+	if err != nil {
+		return err
+	}
+
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	if rowsAffected <= 0 {
+		return api.StatusErrorf(http.StatusNotFound, "DeploymentShape not found")
+	}
+
+	return nil
+}
+
+// RenameDeploymentShape renames a deployment shape.
+func (c *ClusterTx) RenameDeploymentShape(ctx context.Context, deploymentID int64, deploymentShapeID int64, newName string) error {
+	_, err := c.tx.ExecContext(ctx, "UPDATE deployment_shapes SET name=? WHERE deployment_id = ? and id = ?", newName, deploymentID, deploymentShapeID)
+	return err
+}
+
+// DeploymentShapeInstance represents a database deployment shape instance.
+type DeploymentShapeInstance struct {
+	api.DeploymentInstancesPost
+
+	Instance *cluster.Instance
+}
+
+// GetDeploymentShapeInstances returns all instances contained in a DeploymentShape for a given DeploymentShape ID.
+func (c *ClusterTx) GetDeploymentShapeInstances(ctx context.Context, deploymentShapeID int64) ([]*DeploymentShapeInstance, error) {
+	var q *strings.Builder = &strings.Builder{}
+	args := []any{deploymentShapeID}
+
+	q.WriteString(`
+	SELECT
+		s.name AS deployment_shape_name,
+		si.instance_id,
+	FROM deployment_shape_instances AS si
+	INNER JOIN deployment_shapes s ON s.id = si.deployment_shape_id
+	WHERE si.deployment_shape_id = ?
+	`)
+
+	var err error
+	var deploymentShapeInstances []*DeploymentShapeInstance
+
+	err = query.Scan(ctx, c.tx, q.String(), func(scan func(dest ...any) error) error {
+		var deploymentShapeInstance DeploymentShapeInstance
+
+		err := scan(
+			&deploymentShapeInstance.DeploymentShapeName,
+			&deploymentShapeInstance.Instance.ID,
+		)
+		if err != nil {
+			return err
+		}
+
+		deploymentShapeInstances = append(deploymentShapeInstances, &deploymentShapeInstance)
+		return nil
+	}, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	return deploymentShapeInstances, nil
+}
+
+// CreateDeploymentShapeInstance marks an instance as part of a DeploymentShape.
+func (c *ClusterTx) CreateDeploymentShapeInstance(ctx context.Context, deploymentName string, instanceID int64, info api.DeploymentInstancesPost) error {
```

Comment:
This is a code smell to me. The ordering seems incorrect (I think we discussed this before). Why cant we create the entry here after the instance has been created?